### ISSUE TITLE
Unexpected error when change Dc

### ISSUE THREAD

**Issue by zzdhidden** [alik0211/mtproto-core#58]
Step:
* auth
* call messages.getHistory
* call upload.getFile( params, {dcId: doc.dc_id} ) 

This error does not happen usually. Maybe we call `sendEncryptedMessage` before socket open.

```
(node:68903) UnhandledPromiseRejectionWarning: TypeError: Cannot read property 'encrypt' of undefined
    at Socket.obfuscate (/Users/arron/project/mtproto-core/src/transport/obfuscated/index.js:62:28)
    at Socket.send (/Users/arron/project/mtproto-core/src/transport/ws/index.js:55:32)
    at RPC.sendEncryptedMessage (/Users/arron/project/mtproto-core/src/rpc/index.js:633:20)

```

**Comment by alik0211** [comment]
1. What version of the library do you use?
2. Do you run the script in the browser or in node js?

**Comment by zzdhidden** [comment]
@alik0211 git master in node.

**Comment by alik0211** [comment]
@zzdhidden 1. Install library from [npm](https://www.npmjs.com/package/@mtproto/core). What version of node js do you have?

**Comment by zzdhidden** [comment]
node version: v13.0.1

**Comment by alik0211** [comment]
1. Use even versions of node js: 12 or 14
2. Show me all the code so I can repeat the error

**Comment by zzdhidden** [comment]
@alik0211 ok. The frequency of this error is very low. I will reply when I find out the cause of this error.

**Comment by zzdhidden** [comment]
I think I found the problem.
I receive a close event immediately after message event. The `rpc.sendAcks` execute before  rpc reconnect success. 

Error repeat:

`this.emit('close')` after message event.
https://github.com/alik0211/mtproto-core/blob/master/src/transport/socket/index.js#L62

**Comment by zzdhidden** [comment]
Another error appear sometimes.

https://github.com/alik0211/mtproto-core/blob/master/src/rpc/index.js#L539
```
  const waitMessage = this.messagesWaitResponse.get(message.req_msg_id);

      if (message.result._ === 'rpc_error') {
        waitMessage.reject(message.result);
      } else {
        waitMessage.resolve(message.result);
      }
```
The `waitMessage` will be undefined.

**Comment by zzdhidden** [comment]
Unexpected error at https://github.com/alik0211/mtproto-core/blob/master/src/rpc/index.js#L436

`waitMessage.type` error when `waitMessage` undefined.

```
          const waitMessage = self.messagesWaitResponse.get(result.req_msg_id);

          if (!waitMessage) {
            console.warn(
              `RPC result for a non-existent message with id ${result.req_msg_id}`
            );
          }

          const constructorId = this.uint32();
          this.offset -= 4;

          const type =
            constructorId === 558156313 ? 'rpc_error' : waitMessage.type;
```

**Comment by zzdhidden** [comment]
I think all error above occur when socket auto reconnect. The rpc be recreated. The old wait message queue have not reponse because be abandoned.   The new message queue is empty.

**Comment by nodatall** [comment]
I'm having trouble with `waitMessage` being undefined as well. Any ideas for a solution?

**Comment by zzdhidden** [comment]
@nodatall I'm add a timeout for the api before @alik0211 fix it.
`wait(mtproto.call(), 20 * 1000)` or replace `mtproto.call` by yourself.

```
funcion wait(promise, timeout, errorClass = Error, errorMessage = 'timeout') {
    return new Promise((resolve, reject) => {
      let end = false
      let timer
      promise.then(
        (res) => done(false, res),
        (err) => done(true, err)
      )
      if (!end) {
        timer = setTimeout(function () {
          done(true, new errorClass(errorMessage))
        }, timeout)
      }
      function done(rej, res) {
        if (!end) {
          end = true
          if (timer) clearTimeout(timer)
          if (rej) reject(res)
          else resolve(res)
        }
      }
    })
  }
```

**Comment by alik0211** [comment]
Network issues fixed in version 5.2.0 - https://t.me/mtproto_core/25
